### ISSUE TITLE
Add Flower Baseline: Ditto

### ISSUE THREAD

**Issue by oscardilley** [adap/flower#3756]
### Paper

"Ditto: Fair and Robust Federated Learning Through Personalization" by Tian Li, Shengyuan Hu, Ahmad Beirami, Virginia Smith

### Link

https://arxiv.org/abs/2012.04221

### Maybe give motivations about why the paper should be implemented as a baseline.

From what I can see it was proposed for the summer of reproducibility but has not yet been contributed:

https://github.com/adap/flower/issues/2026#issue-1791796534

I have implemented this as a part of other research and would like to contribute the work.

### For first time contributors

- [x]  Read the [`first contribution` doc](https://flower.ai/docs/first-time-contributors.html)
- [x]  Complete the Flower tutorial
- [x]  Read the Flower Baselines docs to get an overview:
    - [x]  [How to use Flower Baselines](https://flower.ai/docs/baselines/how-to-use-baselines.html)
    - [x]  [How to contribute a Flower Baseline](https://flower.ai/docs/baselines/how-to-contribute-baselines.html)

### Prepare - understand the scope

- [x] Read the paper linked above
- [ ] Decide which experiments you'd like to reproduce. The more the better!
- [ ] Follow the steps outlined in [Add a new Flower Baseline](https://flower.ai/docs/baselines/how-to-contribute-baselines.html#add-a-new-flower-baseline).
- [ ] You can use as reference [other baselines](https://github.com/adap/flower/tree/main/baselines) that the community merged following those steps.

### Verify your implementation

- [ ] Follow the steps indicated in the `EXTENDED_README.md` that was created in your baseline directory
- [ ] Ensure your code reproduces the results for the experiments you chose
- [ ] Ensure your `README.md` is ready to be run by someone that is no familiar with your code. Are all step-by-step instructions clear?
- [ ] Ensure running the formatting and typing tests for your baseline runs without errors.
- [ ] Clone your repo on a new directory, follow the guide on your own `README.md` and verify everything runs.

**Comment by jafermarq** [comment]
Hi @oscardilley your are right, Ditto was one of the baselines proposed for Summer of Reproducibility but it wasn't completed. Would you like to add it to Flower Baselines? Do you need any help?

**Comment by oscardilley** [comment]
Hey @jafermarq, yes, I would really like to contribute it as a baseline - I have started reconfiguring it into the required format. I think I am getting on well so far, but is this the best channel to ask any future questions I may have?

**Comment by oscardilley** [comment]
Hi,

The reproduction of Ditto is going well, but I have a couple of sticking
points that I was hoping you could advise on...

From the paper <https://arxiv.org/abs/2012.04221> + repo
<https://github.com/litian96/ditto> does not seem to provide some of
hyperparameters used to achieve the reported results, specifically:

   - Whether the stated learning rate applies to the local and global
   objectives, in relation to Algorithm 2.
   - The number of clients selected for each round of training.
   - The number of local iterations to update the global and personalised
   parameters.

With this in mind, what level of result reproduction do you expect?

Further, would it be acceptable, for the contribution, to only replicate a
subset of the results. i.e. the 'clean' tests without clients acting
maliciously?

Thanks!

**Comment by oscardilley** [comment]
Hi, @jafermarq, please could you advise on this?

**Comment by mercurius80** [comment]
Hi @oscardilley, I am new here but work already in ML Field with flower. I would like to contribute. Do you need any help? Where can I start?
Cheers

**Comment by WilliamLindskog** [comment]
Hi @oscardilley, @mercurius80,

Just checking in here. Are there any updates or is there anything we can do to help?

**Comment by WilliamLindskog** [comment]
It seems that in their repo, they treat the learning rate as one (local and global are the same), I can't find code that distinguishes between them both. 20 clients seem to be default but that they sample 10 of them (fraction 50%). Local iterations seem to be set to 2 and something called `finetune-iters` is 40. You might have already seen this?

**Comment by oscardilley** [comment]
> It seems that in their repo, they treat the learning rate as one (local and global are the same), I can't find code that distinguishes between them both. 20 clients seem to be default but that they sample 10 of them (fraction 50%). Local iterations seem to be set to 2 and something called `finetune-iters` is 40. You might have already seen this?

Hi William, thanks for reaching out. Sorry for the delay on this, I have been busy with work. Good spot on those parameters. Following up from a previous question that I had - for it to be an acceptable reproduction/ baseline, how many/ which of the plots from the paper would it be ideal to reproduce? 

I am hoping to have some time to pick this back up in April.

**Comment by WilliamLindskog** [comment]
@jafermarq any thoughts on how many plots? I'd say the more the merrier, however Figure 4 and 6 should be included I'd say.

**Comment by jafermarq** [comment]
@oscardilley @WilliamLindskog with Baselines we typically aim to have the _core_ results reproduced. In the case of Ditto I see those in Table 1. Having a plot out of these (or even additional experiments is nice) but probably is better to focus on the table. It should be fine having adversaries ratios 20% and 80%. How does that sound ?

**Comment by oscardilley** [comment]
Sounds great, thanks. I will crack on in April time and submit a PR. On 29 Jan 2025, at 17:45, Javier ***@***.***> wrote:﻿
@oscardilley @WilliamLindskog with Baselines we typically aim to have the core results reproduced. In the case of Ditto I see those in Table 1. Having a plot out of these (or even additional experiments is nice) but probably is better to focus on the table. It should be fine having adversaries ratios 20% and 80%. How does that sound ?

**Comment by WilliamLindskog** [comment]
Hi @oscardilley, 

Just checking in here to see if you have the bandwidth to continue with this baseline? 

Best regards
William

**Comment by oscardilley** [comment]
Hi William, I am pretty swamped at the moment. Apologies! Best,Oscar On 22 Apr 2025, at 17:38, William Lindskog ***@***.***> wrote:﻿
  Hi @oscardilley,
Just checking in here to see if you have the bandwidth to continue with this baseline?
Best regards
William—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>



WilliamLindskog left a comment (adap/flower#3756)
Hi @oscardilley,
Just checking in here to see if you have the bandwidth to continue with this baseline?
Best regards
William

**Comment by WilliamLindskog** [comment]
Thanks for letting me know! 

@mercurius80, would you be interested in contributing with this baseline?

**Comment by mercurius80** [comment]
Hi @WilliamLindskog, sure!

**Comment by WilliamLindskog** [comment]
Hi @mercurius80, 

Please have a look at the README [here](https://github.com/adap/flower/tree/main/baselines), and follow the instructions. 

Make sure that it is compatible with the new Flower version, for example the correct use of `clientApp` and `serverApp`. 

Then when you are ready, you can open a PR from your fork, and we will review it, make comments if necessary and help you through the merge process. Something to consider is what results you want/need to reproduce to ensure that the implemented baseline is in line with research study results. 

Best regards
William
